src/graph/network.cc

```diff
@@ -78,104 +91,148 @@ DGL_REGISTER_GLOBAL("network._CAPI_DGLSenderConnect")
+    }
+  });
+
+
+////////////////////////// Distributed Sampler Components ////////////////////////////////
+
+
+DGL_REGISTER_GLOBAL("network._CAPI_SenderSendNodeFlow")
 .set_body([] (DGLArgs args, DGLRetValue* rv) {
     CommunicatorHandle chandle = args[0];
     int recv_id = args[1];
-    // TODO(minjie): could simply use NodeFlow nf = args[2];
     GraphRef g = args[2];
-    const IdArray node_mapping = args[3];
-    const IdArray edge_mapping = args[4];
-    const IdArray layer_offsets = args[5];
-    const IdArray flow_offsets = args[6];
+    const NDArray node_mapping = args[3];
+    const NDArray edge_mapping = args[4];
+    const NDArray layer_offsets = args[5];
+    const NDArray flow_offsets = args[6];
     auto ptr = std::dynamic_pointer_cast<ImmutableGraph>(g.sptr());
     CHECK(ptr) << "only immutable graph is allowed in send/recv";
-    network::Sender* sender = static_cast<network::Sender*>(chandle);
     auto csr = ptr->GetInCSR();
-    // Write control message
-    char* buffer = sender->GetBuffer();
-    *buffer = CONTROL_NODEFLOW;
-    // Serialize nodeflow to data buffer
-    int64_t data_size = network::SerializeSampledSubgraph(
-                             buffer+sizeof(CONTROL_NODEFLOW),
-                             csr,
-                             node_mapping,
-                             edge_mapping,
-                             layer_offsets,
-                             flow_offsets);
-    CHECK_GT(data_size, 0);
-    data_size += sizeof(CONTROL_NODEFLOW);
-    // Send msg via network
-    SendData(sender, buffer, data_size, recv_id);
+    // Create a message for the meta data of ndarray
+    MsgMeta msg(NF_MSG);
+    msg.AddArray(node_mapping);
+    msg.AddArray(edge_mapping);
+    msg.AddArray(layer_offsets);
+    msg.AddArray(flow_offsets);
+    msg.AddArray(csr->indptr());
+    msg.AddArray(csr->indices());
+    msg.AddArray(csr->edge_ids());
+    // First we send meta message
+    int64_t size = 0;
+    char* data = msg.Serialize(&size);
+    network::Sender* sender = static_cast<network::Sender*>(chandle);
+    sender->Send(data, size, recv_id);
+    // Then we send a set of ndarray
+    sender->Send(
+      static_cast<char*>(node_mapping->data),
+      node_mapping.GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(edge_mapping->data),
+      edge_mapping.GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(layer_offsets->data),
+      layer_offsets.GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(flow_offsets->data),
+      flow_offsets.GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(csr->indptr()->data),
+      csr->indptr().GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(csr->indices()->data),
+      csr->indices().GetSize(),
+      recv_id);
+    sender->Send(
+      static_cast<char*>(csr->edge_ids()->data),
+      csr->edge_ids().GetSize(),
+      recv_id);
```

Comment:
Since the `sender` supports multi-threads, how to assure that the receiver receives all these messages in one bundle?

Reply:
> Since the `sender` supports multi-threads, how to assure that the receiver receives all these messages in one bundle?

For sender, multi-thread will be used for sending messages to different remote nodes, for receiver, each thread has an independent message queue for the guarantee of the message order.

Reply:
> I remember we discussed some API like sending a vector of arrays.
> Also, is `Send` non-blocking? how do you manage the memory? Both meta message memory and data in NodeFlow?

I implemented multi-part send into application level. And send is non-blocking, it will push data into a message queue and return immediately. All the message will be simple new/delete in this case.

Reply:
> > I remember we discussed some API like sending a vector of arrays.
> > Also, is `Send` non-blocking? how do you manage the memory? Both meta message memory and data in NodeFlow?
> 
> I implemented multi-part send into application level. And send is non-blocking, it will push data into a message queue and return immediately. All the message will be simple new/delete in this case.

Could you share a pseudo-code of how you achieve it?

Reply:
Here is the place that concerns me: https://github.com/dmlc/dgl/pull/679/files/a8ec9fd31538357ae6726fd73a685bc9d7f0e021#diff-3c750fcd5279ed512c14a3cfa930446cR140
You send data in NodeFlow with multiple sends, which will be sent by multiple threads if the sender has multiple threads. How do you assemble the data in the receiver and construct NodeFlow?

Reply:
> Here is the place that concerns me: https://github.com/dmlc/dgl/pull/679/files/a8ec9fd31538357ae6726fd73a685bc9d7f0e021#diff-3c750fcd5279ed512c14a3cfa930446cR140
> You send data in NodeFlow with multiple sends, which will be sent by multiple threads if the sender has multiple threads. How do you assemble the data in the receiver and construct NodeFlow?

Here is a misunderstanding of my implementation. We send `nodeflow` in multiple send() here, but all the send() will push data into an independent message queue and send it in one thread. Multi-thread send will be only performed for sending data to different nodes with different `recv_id`.  One thread for one queue, and one queue for one node.

Reply:
> > > I remember we discussed some API like sending a vector of arrays.
> > > Also, is `Send` non-blocking? how do you manage the memory? Both meta message memory and data in NodeFlow?
> > 
> > 
> > I implemented multi-part send into application level. And send is non-blocking, it will push data into a message queue and return immediately. All the message will be simple new/delete in this case.
> 
> Could you share a pseudo-code of how you achieve it?

For example, if you want to send a nodeflow to one remote node, you can just do that:

```
send(meta_data)
send(nf.node_mapping)
send(nf.edge_mapping)
send(nf.layer_offset)
send(nf.flow_offset)
...
```

and in receiver:

```
recv(meta_data)
recv(node_mapping)
recv(edge_mapping)
recv(layer_offset)
recv(flow_offset)
```

As we give each socket connection an independent message queue, so you don't need to worry about the order of the message.

Reply:
I think I can see why you want to parallelize communication between destination machines, instead of between messages. If we want to parallelize communication between a pair of machines, we need to create multiple sockets between the two nodes.

However, I'm wondering if it can achieve the maximal throughput of the network. Many graphs have a power-law distribution. it means that communication to different nodes can be very unbalanced, especially in a short period of time. Later on, we need to observe the network throughput. It might be a problem.

Could you please document this in the comment? Both how is parallelization implemented and the potential problem with the current implementation?

Reply:
> I think I can see why you want to parallelize communication between destination machines, instead of between messages. If we want to parallelize communication between a pair of machines, we need to create multiple sockets between the two nodes.
> 
> However, I'm wondering if it can achieve the maximal throughput of the network. Many graphs have a power-law distribution. it means that communication to different nodes can be very unbalanced, especially in a short period of time. Later on, we need to observe the network throughput. It might be a problem.
> 
> Could you please document this in the comment? Both how is parallelization implemented and the potential problem with the current implementation?

Sure. Actually, if we don't need multi-part sending and pack all the data into one message, we can implement multi-thread sending between messages easily. We can leave this potential improvement for the future.